lang/cs/Org.Apache.REEF.Client/YARN/SecurityTokenHelper.cs

```diff
@@ -0,0 +1,124 @@
+// under the License.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Newtonsoft.Json;
+using Org.Apache.REEF.Client.YARN.Parameters;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    /// <summary>
+    /// Helper class for Security token.
+    /// </summary>
+    internal class SecurityTokenHelper
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(SecurityTokenHelper));
+
+        private const string SecurityTokenIdFile = "SecurityTokenId";
+        private const string SecurityTokenPwdFile = "SecurityTokenPwd";
+
+        internal const string DefaultTokenKind = "NULL";
+        internal const string DefaultService = "NULL";
+
+        /// <summary>
+        /// Injectable constructor that accepts a set of serialized tokens.
+        /// </summary>
+        /// <param name="serializedTokenStrings"></param>
+        [Inject]
+        private SecurityTokenHelper([Parameter(typeof(SecurityTokenStrings))] ISet<string> serializedTokenStrings)
+        {
+            if (serializedTokenStrings.Count > 0)
+            {
+                WriteToken(serializedTokenStrings.ToList());
+            }
+            else
+            {
+                TokenKinds = DefaultTokenKind;
+                TokenServices = DefaultService;
+            }
+        }
+
+        internal string TokenKinds { get; private set; }
+
+        internal string TokenServices { get; private set; }
+
+        private static SecurityTokenInfo DeSerializeToken(string serializedTokenStr)
+        {
+            return JsonConvert.DeserializeObject<SecurityTokenInfo>(serializedTokenStr);
+        }
+
+        private void WriteToken(IEnumerable<string> serializedTokenStrings)
+        {
+            IList<byte[]> keys = new List<byte[]>();
+            IList<string> pws = new List<string>();
+            StringBuilder tokenKinds = new StringBuilder();
+            StringBuilder tokenServices = new StringBuilder();
+
+            foreach (var s in serializedTokenStrings)
+            {
+                var token = DeSerializeToken(s);
+                keys.Add(token.SerializedKeyInfoBytes);
+                pws.Add(token.ToeknPassword);
+                if (tokenKinds.Length > 3)
+                {
+                    tokenKinds.Append(":");
+                }
+                tokenKinds.Append(token.TokenKind);
+                tokenKinds.Append(",");
+                tokenKinds.Append(token.SerializedKeyInfoBytes.Length);
+                if (tokenServices.Length > 0)
+                {
+                    tokenServices.Append(":");
+                }
+                tokenServices.Append(token.TokenService);
+            }
+
+            WriteTokens(keys);
+            WritePassword(pws);
+            TokenKinds = tokenKinds.ToString();
+            TokenServices = tokenServices.ToString();
```

Comment:
This is very confusing. a method is called `WriteToken()` (singular!), but it actually does three things: deserialize tokens, write tokens and passwords to disk, and assign new values to the member variables - and all that invoked from the constructor! We need to refactor that piece a bit to make it more readable/maintainable.

I would probably call this method `.ReadTokens()` and let it handle empty input, too, but move the writing elsewhere. Then in the constructor we'll have something like
```C#
if (ReadTokens(serializedTokenStrings))
{
    WriteTokens(DeserializedTokens);
    WritePasswords(DeserializedPasswords);
}
```